File: leetcode/18_4Sum.py
import unittest
from typing import List

class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        nums.sort()
        quadruplets = []
        n = len(nums)
        prev_num_i = None
        for i in range(0, n - 3):
            if nums[i] != prev_num_i:
                prev_num_i = nums[i]
                prev_num_j = None
                for j in range(i + 1, n - 2):
                    if nums[j] != prev_num_j:
                        prev_num_j = nums[j]
                        k, l = j + 1, n - 1
                        prev_num_k, prev_num_l = None, None
                        while k < l:
                            nums_sum = nums[i] + nums[j] + nums[k] + nums[l] 
                            if nums_sum == target:
                                if nums[k] != prev_num_k or nums[l] != prev_num_l:
                                    quadruplets.append([nums[i], nums[j], nums[k], nums[l]])
                                    prev_num_k, prev_num_l = nums[k], nums[l]
                                k += 1
                                l -= 1
                            elif nums_sum > target:
                                l -= 1
                            else:
                                k += 1
        return quadruplets

class Test4Sum(unittest.TestCase):
    def test_1(self):
        self.assertEqual(Solution().fourSum([1, 0, -1, 0, -2, 2], 0), [[-2,-1,1,2],[-2,0,0,2],[-1,0,0,1]])    

    def test_2(self):
        self.assertEqual(Solution().fourSum([2, 2, 2, 2, 2], 8), [[2, 2, 2, 2]])    

    def test_3(self):
        self.assertEqual(Solution().fourSum([2, 2, 2, 2, 2, 2, 2, 2, 2], 8), [[2, 2, 2, 2]])    

File: leetcode/18_4Sum.rs
struct Solution {}

impl Solution {
    pub fn four_sum(nums: Vec<i32>, target: i32) -> Vec<Vec<i32>> {
        let mut nums = nums;
        let n = nums.len();
        let mut quadruplets: Vec<Vec<i32>> = vec![];
        if n < 4 {
            return quadruplets;
        }
        nums.sort();
        for i in 0 .. n - 3 {
            if i == 0 || nums[i] > nums[i - 1]{
                for j in i + 1 .. n - 2 {
                    if j == i + 1 || nums[j] > nums[j - 1] {
                        let mut k  = j + 1;
                        let mut l = n - 1;
                        while k < l {
                            let nums_sum = nums[i] + nums[j] + nums[k] + nums[l];
                            if nums_sum == target {
                                if k == j + 1 || nums[k] > nums[k - 1]{
                                    quadruplets.push(vec![nums[i], nums[j], nums[k], nums[l]]);
                                }
                                k += 1;
                                l -= 1;
                            } else if nums_sum > target {
                                l -= 1;
                            } else {
                                k += 1;
                            }
                        } 
                    } 
                }
            }
        }
        quadruplets
    }
}

#[cfg(test)]
mod four_sum_test {
    use super::*;

    #[test]
    fn test_1() {
        assert_eq!(Solution::four_sum(vec![1,0,-1,0,-2,2], 0), vec![vec![-2,-1,1,2], vec![-2,0,0,2], vec![-1,0,0,1]]);
    }

    #[test]
    fn test_2() {
        assert_eq!(Solution::four_sum(vec![2,2,2,2], 8), vec![vec![2,2,2,2]]);
    }

    #[test]
    fn test_3() {
        assert_eq!(Solution::four_sum(vec![2,2,2,2,2,2,2,2,2,2,2], 8), vec![vec![2,2,2,2]]);
    }
}